CFEP-05: conda-forge policy for providing testing/dev/rc builds on anaconda.org/conda-forge

ericdill opened the pull request:
Thoughts @conda-forge/core ?

ericdill committed:
ENH: First draft of cfep02

msarahan commented:
This is great, @ericdill.  And I'd say you got to it very quickly.  Regarding the point about tooling for how to guide packagers towards how to get on a testing label, I think we can do some kind of detection of extra version information.

For instance, the VersionOrder class in conda (https://github.com/conda/conda/blob/master/conda/version.py#L20) splits things up, and if there's extra text, you end up with a tuple for a given split rather than just the number.  I have used something like this recently to filter conda build's annoying "you should update!" messages: https://github.com/conda/conda-build/blob/a3f7e20e84555884aaa414b81188933f176cb4a0/conda_build/build.py#L495-L503

I think we could use something like that in conda-forge tools somehow to disallow packages with version tags containing strings, and to help set people on the right course of a "testing" package.

ericdill commented:
@msarahan Oh, wow, that's a great solution! Should we consider making the default be uploading to the testing channel unless the `len(VersionOrder(pkg).version[3]) == 1`?

jakirkham commented:
Haven't read this in detail yet, but glad to see you are championing this, @ericdill. Before we dig in too deep, it would be nice to settle how we steer CFEP's forward. Please see PR ( https://github.com/conda-forge/conda-forge-enhancement-proposals/pull/1 ). Once that is settled, would be more than happy to give this a closer look.

jakirkham commented:
Minor note: The testing term in the name confused me a bit as there is a need for testing packages that are built from a stable release, but with new features or in some different way. Maybe we could tweak the title to be more explicit in noting these are pre-releases. Unless of course I'm mistaken and this is trying to resolve both issues. In which case, I would say they are two different things that we should deal with in separate CFEPs. Already have given the testing point a fair bit of thought and am planning to submit a CFEP on that. See this [hackpad](https://conda-forge.hackpad.com/Staged-Releases-DZNKZdgiMbF). Would certainly appreciate your thoughts on it.

ericdill commented:
> Maybe we could tweak the title to be more explicit in noting these are pre-releases

Good idea. I'll update it

> Unless of course I'm mistaken and this is trying to resolve both issues

Nope. You're correct.  I had not even thought of the testing case.  Someone on the call had used the term "testing" and so I am just propagating it here, but this CFEP is really about pre-releases.

> Already have given the testing point a fair bit of thought and am planning to submit a CFEP on that. See this hackpad. Would certainly appreciate your thoughts on it

Awesome! I'll have a look as time permits

ericdill commented:
> Before we dig in too deep, it would be nice to settle how we steer CFEP's forward. Please see PR ( #1 )

Yup. I definitely recognize that those need to get sorted first.  I gave that a read and used the formatting and guidelines from that to inform what sort of general structure/formatting/content cfep02 should have.  I'm definitely putting cfep02 forward with the full understanding that it could need to be totally revamped depending on how #1 goes.

jakirkham commented:
Thanks @ericdill. Was more mentioning the latter point so as to avoid having you wasting cycles on things YTBD. As far as the content goes, it seems like that can develop contemporaneously AFAIAC.

jjhelmus commented:
> @msarahan Oh, wow, that's a great solution! Should we consider making the default be uploading to the testing channel unless the len(VersionOrder(pkg).version[3]) == 1

I do not think we should do this.  Not all packages follow a particular versioning scheme and I think `conda-forge` should be as version scheme agnostic as possible.  I'd rather see uploading to a testing channel be purely opt-in.

pelson commented:
I'd like to see us auto-expire test/dev/rc labelled packages.

Straw-man:

 * testing packages should stick around for 2 weeks
 * dev and rc packages should stick around for 2 months, or as soon as a newer dev/rc package is available (whichever comes first)

pelson commented:
To be completely clear, I'd like to rule "debug" python builds out of scope from this CFEP.

ericdill commented:
I like the idea of auto-expiring the test/dev/rc labeled packages.  Any suggestions as to how this would work?  Where do we store the upload times?

ericdill commented:
(It's on my todo list for this week to circle back to CFEP02)

ericdill commented:
@pelson Where do we track when the package was actually uploaded? I can get the repodata.json for a given label with the following url template: `"https://conda.anaconda.org/{channel}/label/{label}/{platform}/repodata.json"` where channel is "conda-forge", label is "rc" and platform is "linux-64", but the records in repodata.json do not include a timestamp. Here is a code snippet to grab the rc repodata.json and some output to show the keys that they it does contain (which, afaict are not substantially different from the repodata.json for the main label):

```python
import requests
info = requests.get('https://conda.anaconda.org/conda-forge/label/rc/linux-64/repodata.json').json()
info['packages']['matplotlib-2.0.0b1-np111py34_2.tar.bz2']
```

produces

```
{'arch': 'x86_64',
 'binstar': {'channel': 'rc',
  'owner_id': '5528f42ce1dad12974506e8d',
  'package_id': '55f96af4d3234d09d1951089'},
 'build': 'np111py34_2',
 'build_number': 2,
 'depends': ['cycler >=0.10',
  'freetype 2.6*',
  'libpng 1.6*',
  'numpy 1.11*',
  'pyparsing',
  'pyqt',
  'python 3.4*',
  'python-dateutil',
  'pytz',
  'tk'],
 'license': 'PSF-based (http://matplotlib.org/users/license.html)',
 'machine': 'x86_64',
 'md5': 'c261a0140c1ffc169f6b67c78e723b68',
 'name': 'matplotlib',
 'operatingsystem': 'linux',
 'platform': 'linux',
 'requires': [],
 'size': 9806380,
 'subdir': 'linux-64',
 'target-triplet': 'x86_64-any-linux',
 'version': '2.0.0b1'}
```

Anaconda server is clearly tracking this information somewhere since it shows the upload time for all packages that have been uploaded.  Do we know if there is a programmatic way to access this information?

ericdill committed:
Draft 2

ericdill committed:
Disambiguate from "testing" packages

See https://conda-forge.hackpad.com/Staged-Releases-DZNKZdgiMbF for more
discussion and thoughts regarding testing packages.  Also this comment:
https://github.com/conda-forge/conda-forge-enhancement-proposals/pull/3#issuecomment-246134086

ericdill commented:
> I do not think we should do this. Not all packages follow a particular versioning scheme and I think conda-forge should be as version scheme agnostic as possible. I'd rather see uploading to a testing channel be purely opt-in.

Having let this sit for far too long, I am in complete agreement with you @jjhelmus . It seems like it would be a tremendous amount of work to have some sort of mechanism in place to do any sort of reasonable version-scheme validation of the package you're trying to upload and also compare it to the versions that already exist on whatever label you're trying to stick onto your conda package. And after all that work, I do not forsee tangible benefits to such a system that one or two iterations with the package maintainers would not solve.

ericdill commented:
@conda-forge/core could you all give this another read through as time permits? Sorry to let this sit idle for many moons.

Also cc @tacaswell as you expressed interest in ironing out pre-releases

tacaswell commented:
I am not sure of the value of `dev` artifact on conda-forge.  @pelson made a pretty good argument about _not_ dev frequently (that you will end up with so many 'release' if you go to one per commit, that it will become unsustainable both from a disk-space point of view and keeping-track-of-the-things point of view.

I am also not sure how I feel about the aging out policy, if dev versions are dropped re-thinking that might make sense as well.

jakirkham commented:
So PR ( https://github.com/conda-forge/conda-forge-enhancement-proposals/pull/6 ) ended up being called CFEP-02, which we missed. Could we renumber this somehow? It looks like 05 should be ok.

jakirkham commented:
Also sorry if I overlooked this, but did we discuss having dev packages that require other dev packages? How would we like to handle those?

jorisvandenbossche commented:
What is the status on this? Does this need just a decision, or further discussion? 
Anyway I can help? 

A pandas rc release is coming up, so it would be great to be able to use conda-forge to build the release candidate packages.

jakirkham commented:
It needs some tidying as noted above to fix the conflicts.

I guess there is still some discussion on the aging of dev releases. Maybe that could be saved for a CFEP on our retention policy, which we already need anyways.

ericdill committed:
Rename to cfep-05

ericdill commented:
> did we discuss having dev packages that require other dev packages? 

AFAIK this has not been discussed. My vote would be to come up with a policy for how we want to handle `rc` packages and then, if it comes up that we need to be able to support inter-dependent `rc` packages then maybe we could come up with a solution then? It seems like a challenging task to design that correctly and I'd prefer to kick that can down the road a ways.

> @tacaswell says "I am not sure of the value of dev artifact on conda-forge"

I am not particularly opinionated about the value of including "dev" builds in addition to true "rc" builds. As such, I'm happy to defer to those with stronger opinions on the matter. Am I correct in understanding that we only want to support release candidate builds on conda-forge? If so I would propose that this CFEP turns in to "how to make release candidates available to end users" to focus its objective.

------

### So the open questions here are:

1. Do we want to support arbitrary "dev" releases on conda-forge? Or is the scope of this just "release candidate" packages. The semantic difference between dev/rc I guess would be dev == alpha and rc == beta?
1. Do we want to have an automated clean-up policy? Proposal here: 90 days for "rc" packages and then they are moved to "expired-rc" or something. Or just deleted. I have no strong opinions. (The mechanism for how to auto-expire would still need to be worked out)
1. Do we have one label each for "dev" and "rc" (or just "rc" if we rule "dev" out of scope)? Do we have additional labels for each of the packages? i.e., "dev-matplotlib", "rc-pandas", etc... I think that one concern of having all of the dev/rc packages lobbed into common channels is that an unsuspecting user could be attempting to check out the pandas rc and accidentally pull in release candidates for matplotlib/numpy/scipy/etc...
1. Do we want to bypass the labeling mechanism entirely and upload to a different anaconda channel entirely.  Something like "conda-forge-rc". That might make the installation easier since you don't have to include the extra "/label/rc" bit on the end of the channel when you're installing one of these release candidates. Additionally, it keeps all of the release candidates sandboxed away from the non-rc packages. Whether or not that is actually valuable? I'm not so sure.

TomAugspurger commented:
> Do we want to support arbitrary "dev" releases on conda-forge? Or is the scope of this just "release candidate" packages.

From pandas' point of view, I think RC builds would be sufficient

> Do we want to have an automated clean-up policy?

For reference, PyPI keeps these around (e.g. https://pypi.org/project/pandas/0.19.0rc1/), though I don't think conda-forge would *need* to. Once a full release is out, I think that the RC version(s) can be removed (manually, or ideally automatically).

ericdill commented:
Thanks @TomAugspurger . That's a vote from matplotlib and pandas for just doing `rc` and not doing `dev` releases. That seems like enough of the heavyweights from the scientific stack for me to be 100% fine with dropping `dev` releases from the scope of this CFEP. Thoughts @conda-forge/core ?

jjhelmus commented:
> Do we want to support arbitrary "dev" releases on conda-forge? Or is the scope of this just "release candidate" packages.

I agree with @TomAugspurger on this, I think we should limit this to release candidates and similar releases.  Allowing arbitrary "dev" releases could quickly get out of hand, I can see a project wanting to use `conda-forge` for nightly releases or even a release for every commit.  These type of releases have their use and can be very helpful, but I don't think `conda-forge` should be the place for them.

jakirkham commented:
Actually there was someone interested in doing exactly this with dev releases at conda-forge. ( https://github.com/conda-forge/conda-forge.github.io/issues/354 ) @ChrisBarker-NOAA and I had similar thoughts on this that dev releases really don't belong here. The resources that they would require would be truly astronomical if we allowed everyone to try. Nightlies seem well placed under a channel specific to that project. Though tooling like `conda-smithy` could be used to help in that process.

ericdill commented:
Thanks for the input here all. Seems pretty clear that we just want to support release-candidate builds and release builds on the conda-forge channel. With that in mind, the remaining questions are:

So the open questions here are:

1. Do we want to have an automated clean-up policy? Proposal here: 90 days for "rc" packages and then they are moved to "expired-rc" or something. Or just deleted. I have no strong opinions. (The mechanism for how to auto-expire would still need to be worked out)

2. One concern of having all release candidates under the same label is that an unsuspecting user could be attempting to check out the pandas rc and accidentally pull in release candidates for matplotlib/numpy/scipy/etc... One solution to this is to have one label each for "rc" and "rc-<package>" so that, for example, each pandas release candidate gets labeled with "rc" and "rc-pandas". 

3. Do we want to bypass the labeling mechanism entirely and upload to a different anaconda channel entirely. Something like "conda-forge-rc". That might make the installation easier since you don't have to include the extra "/label/rc" bit on the end of the channel when you're installing one of these release candidates. I suppose the only advantage of this is that it keeps all of the release candidates sandboxed away from the release packages. Whether or not that is actually valuable? I'm not so sure. Additionally it does not address the (maybe nonexistent) concern raised in 2.


-----

My thoughts on these three issues are:

1. An automated clean up policy is challenging because AFAICT there is no way to programmatically access the "upload" date for a package on anaconda.org.
2. I am not sure if this is a problem or not. Maybe someone else can weigh in with this one
3. Feels like overengineering a solution to a problem that we don't yet have. My vote is to punt on this and upload to a different anaconda channel if the labeling system is not sufficient.

jakirkham commented:
As for installing RCs selectively, had opened an issue at conda some time back about this.

xref: https://github.com/conda/conda/issues/3125

CJ-Wright commented:
@ericdill Thank you for putting this in! I was thinking about this earlier today in reference to the h5py rc. 
I think the complete graph may be able to help with this, since when we merge a PR we can write to the graph the merge time (via a webservice/feed from the `feedstocks` repo). A script in the bot can follow behind to check which `rc`s expire and expire them.

Additionally the bot could be able to PR `rc` versions to an `rc` branch automatically, which could free up some maintainer resources associated with checking for `rc`s and their proper handling.

dhirschfeld commented:
I'm hitting this problem in https://github.com/conda-forge/staged-recipes/pull/6893 however I think the proposed solution of a `conda-forge/label/rc` label has some problems.

First off, the package I was interested in [`marshmallow`](https://pypi.org/project/marshmallow/#history) has it's first `3.0.0a1` release in Feb 2017. Fast-forward **21** months and we're still only at `3.0.0b20`. `responder` depends on a *beta* version so if I want to use `repsonder` I need to be able to upload beta packages. Sure, you could upload a beta to the rc label, but it seems at that point you may be better off just adding a beta label as well

The much more serious issue IMHO is that if I install `responder` I want to allow conda to install a beta version of `marshmallow` *and only `marshmallow`*. I *really* don't want to be pulling in the dev versions of any other packages which may well completely break my environment

If there is only one `conda-forge/label/rc` label and I add it to my install command then conda will choose to install the latest (dev) versions of all packages which have been uploaded to that single label.

Note: This is the same concern as 2. raised by @ericdill

Until such time as conda gains the ability to specify that a label only be used for a particular package (c.f. https://github.com/conda/conda/issues/3125) then I think the only viable solution is to have individual labels per package.

scopatz commented:
@dhirschfeld - I think it is OK if we make exceptions for the main channel.  But we don't want it to be the norm.

CJ-Wright commented:
I would propose uploading each feedstock's RC packages to its own label, for instance `conda install -c conda-forge/label/rc_matplotlib matplotlib`. This would prevent users from getting other rc packages by mistake.

CJ-Wright commented:
If the packages are in their own feedstock named label do they need to be expired?

dhirschfeld commented:
> *I would propose uploading each feedstock's RC packages to **its own label***

(emphasis mine)

As it stands this is the *only* way to avoid installing dependencies from a pre-release channel. This is the solution we've implemented for our own internal CI/CD but it is a significant amount of work/process to work around a missing feature in conda - https://github.com/conda/conda/issues/9091

ocefpaf commented:
> I would propose uploading each feedstock's RC packages to its own label, for instance `conda install -c conda-forge/label/rc_matplotlib matplotlib`. This would prevent users from getting other rc packages by mistake.

As we discussed in the previous meeting I believe that this is too verbose. As an alternative I propose:

- unified "`dev`" or "`pre`" label, the latter has an advantage of being closer to `pip`;
- recommend the installation with `conda install conda-forge/label/dev::matplotlib`, where the `::` would ensure that only the specified package will come from the `dev` label;
- work with upstream to add a `--pre` option to conda that would do exactly that, making `conda` syntax [closer to `pip`](https://pip.pypa.io/en/stable/reference/pip_install/#install-pre) and making it easier to install pre-releases.

isuruf commented:
Are you sure that 2 works as you suggest? My experience has been that some other packages from that channel will be chosen as well.

ocefpaf commented:
> Are you sure that 2 works as you suggest? My experience has been that some other packages from that channel will be chosen as well.

I never found a case where that did not work, if you find one please report here so we can check if the problem is with `conda` or the package metadata. In the gist below there are two examples, one for `gmt` and another one for `matplotlib`. Note that in the `matplotlib` example `conda` correctly pulled `matplotlib-base` from `rc` as well.

https://gist.github.com/ocefpaf/b22eaae367c83dbb9375bd3e80955f9f

CJ-Wright commented:
@conda-forge/core I don't think we're ready to call a vote yet, but it would be nice to get some rough ideas on which methods are preferred. Note that multiple approaches could be taken, especially if they have different time windows

1. unified "dev" or "pre" label, the latter has an advantage of being closer to pip
1. recommend the installation with `conda install conda-forge/label/dev::matplotlib`, where the `::` would ensure that only the specified package will come from the `dev` label
1. work with upstream to add a `--pre` option to conda that would do exactly that, making conda syntax closer to pip and making it easier to install pre-releases.
1. put each package's pre/rc into its own channel `conda install -c conda-forge/label/rc_matplotlib matplotlib`

If possible please vote with reactions:
1. 😄 
2. 🚀  
3. 👀 
4. 🎉 

(I tried to pick the reactions without bias, but there are too few)

ocefpaf commented:
Note that `2` can happen now. We should recommend that approach to install from a `label` regardless of the other options. As far as I know there are no side effects, only advantages.

1-2-3 can also happen together.

4 is probably easier to implement than 1-2-3 but leads to more typing fell less like `pip pre`.

CJ-Wright commented:
Ok, based on the emoji vote, I'll rewrite this proposal to reflect method 2 for now.

jjhelmus commented:
It would be prudent to test if `conda install conda-forge/label/dev::matplotlib` really only installs `matplotlib` from the `dev` channel.  From my limited testing other packages in the `dev` channel will also be installed in some cases.

CJ-Wright commented:
Is there a way to get all the dev packages in existence?

jjhelmus commented:
https://anaconda.org/conda-forge/repo?type=conda&label=dev

isuruf commented:
I added a `dev` label to scipy to test some things. I can't get it to work with strict channel priority

1. `conda create -n test11 "conda-forge/label/dev::scipy" --dry-run` works.
2. `conda create -n test11 "conda-forge/label/dev::scipy" --dry-run --strict-channel-priority` leads to unsolvable

CJ-Wright commented:
Interesting! Thank you @isuruf!

isuruf commented:
`conda create -n test11 "conda-forge/label/dev::scipy" "conda-forge/label/dev::numpy" --dry-run` doesn't give me numpy from `dev` label.

CJ-Wright commented:
Is it possible that dev `scipy` wasn't built with dev `numpy` and that they are conflicting?

isuruf commented:
> Is it possible that dev scipy wasn't built with dev numpy and that they are conflicting?

I don't understand. Why should what scipy was built with matter here?

CJ-Wright commented:
Does the dev scipy require dev numpy versions?

Or maybe better stated, is it possible that the numpy and scipy builds are incompatible?

jasongrout commented:
I just tried this syntax on macOS:

```sh
conda create -n del-jlabwdev "conda-forge/label/prerelease-jupyterlab::jupyterlab=1.2.0a1"
```
I had a lot of conflicts. On the other hand, doing 
```sh
conda create -n del-jlabwdev -c conda-forge/label/prerelease-jupyterlab jupyterlab=1.2.0a1
```
worked fine.

My condarc is:

```
channels:
  - conda-forge
show_channel_urls: true
channel_priority: strict
auto_activate_base: false
```
Did I do something wrong with the `::` syntax?

CJ-Wright commented:
Can you post the results of those two installations?

jasongrout commented:
```
~% conda --version
conda 4.7.12
~% conda create -n del-test1 "conda-forge/label/prerelease-jupyterlab::jupyterlab=1.2.0a1"

```
<details><summary>Output</summary>

```
Collecting package metadata (current_repodata.json): done
Solving environment: failed with repodata from current_repodata.json, will retry with next repodata source.
Collecting package metadata (repodata.json): done
Solving environment: \ 
Found conflicts! Looking for incompatible packages.
This can take several minutes.  Press CTRL-C to abort.
failed                                                                                                                                                                                   

UnsatisfiableError: The following specifications were found to be incompatible with each other:



Package ssl_match_hostname conflicts for:
tornado[version='!=6.0.0,!=6.0.1,!=6.0.2'] -> ssl_match_hostname
Package sqlite conflicts for:
python[version='>=3.5'] -> sqlite[version='3.13.*|3.20.*|3.9.*|>=3.24.0,<4.0a0|>=3.25.1,<4.0a0|>=3.25.2,<4.0a0|>=3.25.3,<4.0a0|>=3.26.0,<4.0a0|>=3.28.0,<4.0a0']
Package libcxx conflicts for:
python[version='>=3.5'] -> libcxx[version='>=4.0.1']
Package libffi conflicts for:
python[version='>=3.5'] -> libffi[version='>=3.2.1,<3.3.0a0']
Package nbformat conflicts for:
notebook[version='>=4.3.1'] -> nbformat
Package jsonschema conflicts for:
jupyterlab_server[version='>=1.0.0,<2.0.0'] -> jsonschema[version='>=3.0.1']
Package readline conflicts for:
python[version='>=3.5'] -> readline[version='6.2.*|7.0|>=7.0,<8.0a0|>=8.0,<9.0a0']
Package certifi conflicts for:
tornado[version='!=6.0.0,!=6.0.1,!=6.0.2'] -> certifi
Package tornado conflicts for:
notebook[version='>=4.3.1'] -> tornado[version='>=4.1,<7|>=4|>=4,<6|>=5.0,<7']
Package json5 conflicts for:
jupyterlab_server[version='>=1.0.0,<2.0.0'] -> json5
Package markupsafe conflicts for:
jinja2[version='>=2.10'] -> markupsafe[version='>=0.23']
Package pip conflicts for:
python[version='>=3.5'] -> pip
Package jinja2 conflicts for:
jupyterlab_server[version='>=1.0.0,<2.0.0'] -> jinja2[version='>2.10*']
notebook[version='>=4.3.1'] -> jinja2
Package xz conflicts for:
python[version='>=3.5'] -> xz[version='5.0.*|5.2.*|>=5.2.3,<5.3.0a0|>=5.2.4,<5.3.0a0']
Package terminado conflicts for:
notebook[version='>=4.3.1'] -> terminado[version='>=0.8.1']
Package backports_abc conflicts for:
tornado[version='!=6.0.0,!=6.0.1,!=6.0.2'] -> backports_abc[version='>=0.4']
Package nbconvert conflicts for:
notebook[version='>=4.3.1'] -> nbconvert
Package tk conflicts for:
python[version='>=3.5'] -> tk[version='8.5.*|8.6.*|>=8.6.8,<8.7.0a0|>=8.6.9,<8.7.0a0']
Package zlib conflicts for:
python[version='>=3.5'] -> zlib[version='1.2.*|1.2.11|1.2.8|>=1.2.11,<1.3.0a0']
Package ipaddress conflicts for:
notebook[version='>=4.3.1'] -> ipaddress
Package jupyter_client conflicts for:
notebook[version='>=4.3.1'] -> jupyter_client[version='>=5.2.0|>=5.3.1']
Package jupyter_core conflicts for:
notebook[version='>=4.3.1'] -> jupyter_core[version='>=4.4.0']
Package ncurses conflicts for:
python[version='>=3.5'] -> ncurses[version='>=6.1,<6.2.0a0']
Package send2trash conflicts for:
notebook[version='>=4.3.1'] -> send2trash
Package singledispatch conflicts for:
tornado[version='!=6.0.0,!=6.0.1,!=6.0.2'] -> singledispatch
Package notebook conflicts for:
jupyterlab_server[version='>=1.0.0,<2.0.0'] -> notebook[version='>=4.2.0']
Package setuptools conflicts for:
jinja2[version='>=2.10'] -> setuptools
Package futures conflicts for:
tornado[version='!=6.0.0,!=6.0.1,!=6.0.2'] -> futures
Package pyzmq conflicts for:
notebook[version='>=4.3.1'] -> pyzmq[version='>=17']
Package traitlets conflicts for:
notebook[version='>=4.3.1'] -> traitlets[version='>=4.2.1|>=4.3']
Package prometheus_client conflicts for:
notebook[version='>=4.3.1'] -> prometheus_client
Package bzip2 conflicts for:
python[version='>=3.5'] -> bzip2[version='>=1.0.6,<2.0a0']
Package openssl conflicts for:
python[version='>=3.5'] -> openssl[version='1.0.*|>=1.0.2o,<1.0.3a|>=1.0.2p,<1.0.3a|>=1.1.1a,<1.1.2a']
Package ipython_genutils conflicts for:
notebook[version='>=4.3.1'] -> ipython_genutils
Package ipykernel conflicts for:
notebook[version='>=4.3.1'] -> ipykernel
Note that strict channel priority may have removed packages required for satisfiability.

```
</details>

```
~% conda create -n del-test3 -c conda-forge/label/prerelease-jupyterlab jupyterlab=1.2.0a1
```

<details><summary>Output</summary>

```
Collecting package metadata (current_repodata.json): done
Solving environment: done

## Package Plan ##

  environment location: /Users/jgrout6/miniconda3/envs/del-test3

  added / updated specs:
    - jupyterlab=1.2.0a1


The following NEW packages will be INSTALLED:

  appnope            conda-forge/osx-64::appnope-0.1.0-py37_1000
  attrs              conda-forge/noarch::attrs-19.2.0-py_0
  backcall           conda-forge/noarch::backcall-0.1.0-py_0
  bleach             conda-forge/noarch::bleach-3.1.0-py_0
  bzip2              conda-forge/osx-64::bzip2-1.0.8-h01d97ff_1
  ca-certificates    conda-forge/osx-64::ca-certificates-2019.9.11-hecc5488_0
  certifi            conda-forge/osx-64::certifi-2019.9.11-py37_0
  decorator          conda-forge/noarch::decorator-4.4.0-py_0
  defusedxml         conda-forge/noarch::defusedxml-0.6.0-py_0
  entrypoints        conda-forge/osx-64::entrypoints-0.3-py37_1000
  importlib_metadata conda-forge/osx-64::importlib_metadata-0.23-py37_0
  ipykernel          conda-forge/osx-64::ipykernel-5.1.2-py37h5ca1d4c_0
  ipython            conda-forge/osx-64::ipython-7.8.0-py37h5ca1d4c_0
  ipython_genutils   conda-forge/noarch::ipython_genutils-0.2.0-py_1
  jedi               conda-forge/osx-64::jedi-0.15.1-py37_0
  jinja2             conda-forge/noarch::jinja2-2.10.3-py_0
  json5              conda-forge/noarch::json5-0.8.5-py_0
  jsonschema         conda-forge/osx-64::jsonschema-3.1.1-py37_0
  jupyter_client     conda-forge/osx-64::jupyter_client-5.3.3-py37_1
  jupyter_core       conda-forge/noarch::jupyter_core-4.5.0-py_0
  jupyterlab         conda-forge/label/prerelease-jupyterlab/noarch::jupyterlab-1.2.0a1-py_0
  jupyterlab_server  conda-forge/noarch::jupyterlab_server-1.0.6-py_0
  libcxx             conda-forge/osx-64::libcxx-9.0.0-0
  libcxxabi          conda-forge/osx-64::libcxxabi-9.0.0-0
  libffi             conda-forge/osx-64::libffi-3.2.1-h6de7cb9_1006
  libsodium          conda-forge/osx-64::libsodium-1.0.17-h01d97ff_0
  markupsafe         conda-forge/osx-64::markupsafe-1.1.1-py37h1de35cc_0
  mistune            conda-forge/osx-64::mistune-0.8.4-py37h1de35cc_1000
  more-itertools     conda-forge/noarch::more-itertools-7.2.0-py_0
  nbconvert          conda-forge/osx-64::nbconvert-5.6.0-py37_1
  nbformat           conda-forge/noarch::nbformat-4.4.0-py_1
  ncurses            conda-forge/osx-64::ncurses-6.1-h0a44026_1002
  notebook           conda-forge/osx-64::notebook-6.0.1-py37_0
  openssl            conda-forge/osx-64::openssl-1.1.1c-h01d97ff_0
  pandoc             conda-forge/osx-64::pandoc-2.7.3-0
  pandocfilters      conda-forge/noarch::pandocfilters-1.4.2-py_1
  parso              conda-forge/noarch::parso-0.5.1-py_0
  pexpect            conda-forge/osx-64::pexpect-4.7.0-py37_0
  pickleshare        conda-forge/osx-64::pickleshare-0.7.5-py37_1000
  pip                conda-forge/osx-64::pip-19.2.3-py37_0
  prometheus_client  conda-forge/noarch::prometheus_client-0.7.1-py_0
  prompt_toolkit     conda-forge/noarch::prompt_toolkit-2.0.10-py_0
  ptyprocess         conda-forge/noarch::ptyprocess-0.6.0-py_1001
  pygments           conda-forge/noarch::pygments-2.4.2-py_0
  pyrsistent         conda-forge/osx-64::pyrsistent-0.15.4-py37h01d97ff_0
  python             conda-forge/osx-64::python-3.7.3-h93065d6_1
  python-dateutil    conda-forge/noarch::python-dateutil-2.8.0-py_0
  pyzmq              conda-forge/osx-64::pyzmq-18.1.0-py37hee98d25_0
  readline           conda-forge/osx-64::readline-8.0-hcfe32e1_0
  send2trash         conda-forge/noarch::send2trash-1.5.0-py_0
  setuptools         conda-forge/osx-64::setuptools-41.4.0-py37_0
  six                conda-forge/osx-64::six-1.12.0-py37_1000
  sqlite             conda-forge/osx-64::sqlite-3.30.1-h93121df_0
  terminado          conda-forge/osx-64::terminado-0.8.2-py37_0
  testpath           conda-forge/noarch::testpath-0.4.2-py_1001
  tk                 conda-forge/osx-64::tk-8.6.9-h2573ce8_1003
  tornado            conda-forge/osx-64::tornado-6.0.3-py37h01d97ff_0
  traitlets          conda-forge/osx-64::traitlets-4.3.3-py37_0
  wcwidth            conda-forge/noarch::wcwidth-0.1.7-py_1
  webencodings       conda-forge/noarch::webencodings-0.5.1-py_1
  wheel              conda-forge/osx-64::wheel-0.33.6-py37_0
  xz                 conda-forge/osx-64::xz-5.2.4-h1de35cc_1001
  zeromq             conda-forge/osx-64::zeromq-4.3.2-h6de7cb9_2
  zipp               conda-forge/noarch::zipp-0.6.0-py_0
  zlib               conda-forge/osx-64::zlib-1.2.11-h0b31af3_1006


Proceed ([y]/n)? y

Preparing transaction: done
Verifying transaction: done
Executing transaction: done
#
# To activate this environment, use
#
#     $ conda activate del-test3
#
# To deactivate an active environment, use
#
#     $ conda deactivate

```

</details>

CJ-Wright committed:
update to rc_package_name

CJ-Wright commented:
I've updated this cfep to reflect our core meeting this week. I think there are two main outcomes from that meeting:
1. The only approach that will consistently produce usable solves and environments that support dev/rc packages is to have independent dev/rc labels for each package (eg. `.../rc_matplotlib`).
2. The user experience of this is not as easy as pip's `--pre` flag. With this in mind we can build out provisional dev/rc support (this CFEP) and make decisions about future work around the user experience once we have established demand and how it is being used.

CJ-Wright commented:
@conda-forge/core
This PR falls under the CFEP policy, please vote and/or comment on this PR.
This PR needs 60% (11) of core to vote yea to pass.
To vote please leave Approve (yea) or Request Changes (nay) reviews.
If you would like changes to the current language please leave a comment or push to this branch.
This vote will end on Oct 25th (1 week).

jasongrout commented on cfep-05.md:
```diff
@@ -0,0 +1,136 @@
+There are a number of recipe maintainers who have asked for guidance regarding
+how to make pre-release builds available with the CI infrastructure on
+conda-forge. This document outlines how those builds should be labeled on
+anaconda.org so that the more adventurous users can try our these pre-release
```
```suggestion
anaconda.org so that the more adventurous users can try out these pre-release
```

isuruf commented on cfep-05.md:
```diff
@@ -0,0 +1,136 @@
+...
+channels:
+  targets:
+    - [conda-forge, rc_matplotlib]
```
This syntax is wrong.

CJ-Wright replied:
What is the correct syntax? I modified the existing matplotlib development branch https://github.com/conda-forge/matplotlib-feedstock/blob/devel/conda-forge.yml

isuruf replied:
See point 4 at https://github.com/conda-forge/conda-smithy#making-a-new-feedstock

There are 2 files.
1. `recipe/conda_build_config.yaml` - this is what actually works
2. `conda-forge.yml` - this was the conda-smithy <3 way, but this was partially removed. This is only used for the README now. Ideally someone should fix the README rendering to use `recipe/conda_build_config.yaml`

CJ-Wright replied:
Thank you!

CJ-Wright committed:
typo

mbargull commented on cfep-05.md:
```diff
@@ -0,0 +1,136 @@
+<tr><td> Status </td><td> Proposed </td></tr>
+<tr><td> Author(s) </td><td> Eric Dill &lt;thedizzle@gmail.com&gt;</td></tr>
+<tr><td> Created </td><td> Sept 09, 2016</td></tr>
+<tr><td> Updated </td><td> Sept 09, 2016</td></tr>
```
`datetime.today().strftime("%b %d, %Y")` ;)

mbargull replied:
```suggestion
<tr><td> Updated </td><td> Oct 18, 2019</td></tr>
```

mbargull commented on cfep-05.md:
```diff
@@ -0,0 +1,136 @@
+
+## Motivation
+
+Pypi allows package maintainers to upload "prerelease" builds which users can
```
```suggestion
PyPI allows package maintainers to upload "prerelease" builds which users can
```

CJ-Wright committed:
fix syntax for labels

mbargull commented on cfep-05.md:
````diff
@@ -0,0 +1,136 @@
+are still manageable.  The analogous command with `conda` is (for example):
+
+```
+conda install -c conda-forge/label/rc_matplotlib matplotlib
````
What are anaconda.org's restrictions on valid labels?

1. Are all valid package name characters valid in labels?
2. Is `/` valid, too?

We have to get confirmation for 1. from Anaconda before giving our blessing.

If 2. is possible too, I would like to propose `rc/matplotlib` as the label. It would be more readable and avoid any future ambiguities.

jasongrout replied:
+1 to `conda-forge/label/rc/matplotlib` if that is possible, rather than mangling the name of the package.

CJ-Wright replied:
I'm worried that this might imply that there is an `rc` channel with a `matplotlib` channel inside it, which is what we want to avoid if possible. @msarahan @soapy1 thoughts?

mbargull replied:
> I'm worried that this might imply that there is an `rc` channel with a `matplotlib` channel inside it

I don't share that worry. However, anaconda.org does not seem to like it:
You can create such label, but you get a 404 for the `repodata.json`. See
https://anaconda.org/mbargull/_label-test
https://conda.anaconda.org/mbargull/label/rc/_label-test/noarch/repodata.json

I tried out some other characters that the anaconda.org web GUI spits out as valid. A single `:` seems to work, `::` not.

https://conda.anaconda.org/mbargull/label/rc:_label-test/noarch/repodata.json
https://conda.anaconda.org/mbargull/label/rc::_label-test/noarch/repodata.json
```sh
> CONDA_DEFAULT_CHANNELS= conda search 'mbargull/label/rc::_label-test::_label-test'
Loading channels: done
No match found for: mbargull/label/rc::_label-test::_label-test. Search: mbargull/label/rc::_label-test::*_label-test*

PackagesNotFoundError: The following packages are missing from the target environment:
  - mbargull/label/rc::_label-test::_label-test


> CONDA_DEFAULT_CHANNELS= conda search 'mbargull/label/rc:_label-test::_label-test'
Loading channels: done
# Name                       Version           Build  Channel             
_label-test                    0.1.0               0  mbargull/label/rc:_label-test
```

This would mean `rc:matplotlib` or `matplotlib:rc` would be possible. IDK if this would confuse users of the `channel::package` notation, though. And I don't know if this would be officially supported.

We need someone from Anaconda to comment on this. My vote would be to make anaconda.org support `/` in labels, e.g., `rc/matplotlib` or `matplotlib/rc`.

mbargull commented on cfep-05.md:
````diff
@@ -0,0 +1,135 @@
+pip install PACKAGE_NAME --pre
+```
+
+The same semantics are slightly more cumbersome within the `conda` ecosystem but
````
```suggestion
The same semantics are slightly more cumbersome within the Conda ecosystem but
```

mbargull commented on cfep-05.md:
```diff
@@ -0,0 +1,135 @@
+relevant labels that package maintainers have affixed to their pre-release
+packages. These are "pre" (1), "prerelease" (2), "rc" (2) and "test" (1).  The
+numbers in parentheses represent the number of projects which have released
+these pre-release packages.  With six total projects using non-main labels, I do
```
Do we need to update those numbers? (Or just reflect that those are from when the proposal has been created?)

jasongrout commented:
Is there some reasoning behind using `rc_matplotlib` vs `matplotlib_rc`? Having `rc` first calls attention to it, but usually we talk about version numbers in the order `package` `version`.

jasongrout commented:
Also, written out as most people would use it, `conda_forge/label/rc_maplotlib`, the `rc` can get kind of lost in the middle compared to `conda_forge/label/matplotlib_rc`

isuruf commented on cfep-05.md:
````diff
@@ -0,0 +1,135 @@
+```yaml
+...
+channel_targets:
+  - conda-forge rc_matplotlib
````
Can you also add an example for how to use a dev package in another dev package?

CJ-Wright replied:
Is this something we want to support? Would users get the correct env if they don't have both dev channels in their .condarc/cli call?

isuruf replied:
They won't, but the user should be given specific instructions either way. I think we should support it or say that we don't.

CJ-Wright replied:
I would lean toward not supporting this for now, seeing if people need it and then revisiting. This way we are more likely to provide a stable system day one.

isuruf replied:
I'd need it and I saw some people asking about it in gitter, but can't find it right now.

CJ-Wright replied:
Can you please push up the language you want to see?

CJ-Wright replied:
We also need to make sure that these extra channels don't leak into downstream `main` builds. Maybe add a poison pill to the build system (unless the linter can see if you PRed into master) that doesn't build on master if you pull from `rc/dev`.

isuruf replied:
Shall I add a section like `Work Items`?

CJ-Wright replied:
That would be great! Thank you!

isuruf replied:
Have a look and feel free to change it.

mbargull commented on cfep-05.md:
```diff
@@ -0,0 +1,135 @@
+
+## Specification
+
+Mark the pre-release package on anaconda.org as "dev" or "rc". The semantics of
```
For the TLDR; readers the `Mark [...] as "dev" or "rc"` may sound like labelling as `dev`/`rc` (full name, not affix).

CJ-Wright commented:
@jasongrout I don't have strong opinions one way or the other. Although I think some packages are named `package-name-dev`, which could cause some problems for a dev/rc last naming convention.

mbargull commented on cfep-05.md:
```diff
@@ -0,0 +1,135 @@
+
+So make sure that you ***tag*** your package in such a way that the package name
+that conda-build spits out will sort the package uploaded with an "rc" label
+higher than the package uploaded with the "dev" label.
```
Do we need this? This can be mentioned informally, but should not be part of the specification.
It might also confuse people that deal with packages with different prerelease version schemes, e.g., GCC.

jasongrout commented:
Hmm, so `conda_forge/label/dev_package_name_dev` vs `conda_forge/label/package_name_dev_dev`? Would those `-dev` packages have dev releases themselves?

I think I prefer the rc/dev at the end for the two reasons I mention above: it reads more naturally (even your package name examples above indicate it reads more naturally at the end :), and it stands out more at the end rather than being lost in the middle of `conda_forge/label/rc_matplotlib`.

mbargull commented on cfep-05.md:
````diff
@@ -0,0 +1,135 @@
+to install and replace `LABEL` with "rc" or "dev":
+
+```
+conda install -c conda-forge -c conda-forge/label/LABEL_PACKAGE_NAME PACKAGE_NAME
````
I would assume that the user will usually want to give the prerelease channel a higher priority, no?
```suggestion
conda install -c conda-forge/label/LABEL_PACKAGE_NAME PACKAGE_NAME -c conda-forge
```

mbargull commented on cfep-05.md:
````diff
@@ -0,0 +1,135 @@
+For example, let's install matplotlib from the "rc" label:
+
+```
+conda install -c conda-forge -c conda-forge/label/rc_matplotlib matplotlib
````
I would assume that the user will usually want to give the prerelease channel a higher priority, no?
```suggestion
conda install -c conda-forge/label/rc_matplotlib matplotlib -c conda-forge
```

CJ-Wright commented:
@mbargull would you mind pushing your updates to the repo directly? I can't seem to accept the suggestions.

mbargull commented:
Weird. Sure can do. Are all suggestions appropriate?

CJ-Wright commented:
I think so.

mbargull committed:
update date, fix typos, change channel order

mbargull commented:
> I can't seem to accept the suggestions.

FWIW, it didn't work for me either via the web GUI..

isuruf committed:
Add channel_sources and rerender note

isuruf committed:
Add work items

dhirschfeld commented on cfep-05.md:
```diff
@@ -0,0 +1,164 @@
+
+To create a `dev` or `rc` package a PR can be issued into the `dev` or `rc` branch of the
+feedstock.
+This branch must change the `recipe/conda_build_config.yaml` to point to the `dev_<package_name>` or `rc_<package_name>`.
```
This sentence is missing the name of what `dev_<package_name>` actually is - is it a channel, a label, a package name, or something else??

I assume it's talking about a label in which case it should read:
```
This branch must change the `recipe/conda_build_config.yaml` to include either the `dev_<package_name>` or `rc_<package_name>` label.
```

If that's not correct, it should be reworded to be unambiguous what it is talking about.

CJ-Wright committed:
updates from comments

CJ-Wright commented:
I think this is now mostly awaiting agreement on the label syntax. My understanding of the current contenders are:
1. `.../conda-forge/label/rc_matplotlib`
2. `.../conda-forge/label/matplotlib_rc`
3. `.../conda-forge/label/rc/matplotlib` (This is not currently supported by ~~conda~~ anaconda.org)

Please vote via emojis:
1. 🎉 
2. 👀 
3. 🚀

h-vetinari commented on cfep-05.md:
```diff
@@ -0,0 +1,166 @@
+
+## FAQ
+
+* What about "alpha" and "beta" labels?
+
+There is not yet heavy use of labels on the conda-forge channel on anaconda.org.
+Given the light usage of labels thus far it seems rather unnecessary to
+introduce many labels.  "dev" and "rc" seem like a nice compromise.
```
Just FYI for those who haven't seen it - there's two competing PEPs for increasing the release cadence of python. One of them (see PEP text and discussion [here](https://discuss.python.org/t/pep-605-a-rolling-feature-release-stream-for-cpython/2418)) is proposing to add a rolling release stream with alphas (ABI-breaking) and betas (stable).

The PEP is not accepted yet, and this CFEP could of course be changed later if necessary, so I don't think there's anything to do, I just wanted to point this out, because it touches on this topic.

mbargull replied:
Yes, we could change the policy later if needed. Personally, I'd assume future CPython alphas could go into `python_dev` and betas into `python_dev` fitting nicely with this proposal.

But basically, I just wanted to say thanks for the reference! (Though, TBH, I TLDR;'d that one, haha...)

h-vetinari replied:
Actually, it was the other [PEP](https://www.python.org/dev/peps/pep-0602/) that ended up being accepted. This means a smaller change compared to the status quo, just a faster release cadence.

h-vetinari reviewed (commented):
Just mentioning PEP 605.

jakirkham commented:
Has anyone thought about split package scenarios? Right now we are discussing labels that have the package name. Though it is less clear (at least to me) what we do when there are several packages being published.

CJ-Wright commented:
I think the labels should be the feedstock names (for example `matplotlib`). I expect the release cadence for all the packages in a split-package feedstock to be the same.

jakirkham commented:
What about new functionality that is being developed out-of-tree until it stabilizes?

mbargull commented:
> What about new functionality that is being developed out-of-tree until it stabilizes?

Can you clarify? Are you hinting at creating more diverse labels than `*_dev` and `*_rc`? If yes, I don't think we necessarily need to incorporate that into this proposal. If some projects needed more than dev/rc for that, it might make more sense to do that outside of the `conda-forge` channels.

mbargull commented:
Due to technical limitations (re. the `/` separator) or technical concerns (re. `::` as a separator), we have
- `pkg_{dev,rc}`
- `{dev,rc}_pkg`

as the current viable options for this proposal. Please "emoji vote" on https://github.com/conda-forge/conda-forge-enhancement-proposals/pull/3#issuecomment-545011203 .
Apart from choosing from those options, is there anything else to stopping this CFEP from being voted on?

CJ-Wright commented:
The Eyes have it, the Eyes have it! I'll update the channel name accordingly.

mbargull commented:
> I think the labels should be the feedstock names (for example `matplotlib`). I expect the release cadence for all the packages in a split-package feedstock to be the same.

Could we either add this or the opposing option to the CFEP to get a clear recommendation?
(cc @mariusvniekerk, @dopplershift)
xref: https://gitter.im/conda-forge/conda-forge.github.io?at=5dde30c4b065c6433cf2f0eb

ericdill commented:
Adding a note from a recent dev meeting: We should implement this in whatever way is as simple as possible for the conda-forge tooling ecosystem. If there is enough of sentiment that people want something more sophisticated, PRs welcome.

CJ-Wright commented:
@conda-forge/core
This PR falls under the CFEP policy, please vote and/or comment on this PR.
This PR needs 60% (12) of core to vote yea to pass.
To vote please leave Approve (yea) or Request Changes (nay) reviews.
If you would like changes to the current language please leave a comment or push to this branch.
This vote will end on Jan 22nd (2 weeks).

CJ-Wright commented:
@ericdill since you opened this PR you can't approve it via the review. Are you on board with this?

jasongrout commented:
>  This vote will end on Jan 22nd (2 weeks).

This is just a ping that the vote ends tomorrow. As a person in the community, I'm happy that this proposal is under consideration.

CJ-Wright commented:
This vote passes!

CJ-Wright commented on cfep-05.md:
```diff
@@ -0,0 +1,166 @@
+
+<table>
+<tr><td> Title </td><td> Providing pre-release builds on the conda-forge anaconda channel </td>
+<tr><td> Status </td><td> Proposed </td></tr>
```
```suggestion
<tr><td> Status </td><td> Accepted </td></tr>
```

CJ-Wright commented:
@ericdill if you could please merge my change to mark this as accepted I'll merge this.

scopatz commented:
Let's fix in repo

jasongrout commented:
In https://github.com/conda-forge/cfep/pull/3#issuecomment-555629653, @CJ-Wright indicated that the CFEP would be updated to the naming convention `.../conda-forge/label/matplotlib_rc`, but it seems the text was never updated, so what was merged was the label convention `rc_matplotlib`. Was this intentional? (I don't know if there were other conversations off this thread that decided the `rc_package` convention instead of `package_rc`.)

CJ-Wright commented:
Sorry my fault, I'll fix that in a follow up pr

jasongrout commented:
> Sorry my fault, I'll fix that in a follow up pr

I opened https://github.com/conda-forge/cfep/pull/21 to update the CFEP 5 text. I'm not sure exactly the right procedure for doing this, so please let me know if something is amiss.
